Docs: Refactor "Environment variables" into 3 articles (Diátaxis)

benjaoming opened the pull request:
This breaks up our current Environment Variables guide:

* Avoids mix of "pre-defined" environment reference with "user-defined" scope
* Replaces "user-defined" with "custom"
* Adds more guidance in howto + a screenshot
* Discusses alternatives to non-secret environment variables

Refs: #9746 

<!-- readthedocs-preview docs start -->
---
:books: Documentation previews :books:

- User's documentation (`docs`): https://docs--9966.org.readthedocs.build/en/9966/

<!-- readthedocs-preview docs end -->

<!-- readthedocs-preview dev start -->
- Developer's documentation (`dev`): https://dev--9966.org.readthedocs.build/en/9966/

<!-- readthedocs-preview dev end -->

benjaoming commented:
@ericholscher I dare say that this is ready for review :+1:

ericholscher reviewed (commented):
I'm a little confused on the differences in these 2 pages. I think a better title on `Configuring environment variables` would help. It's not about configuring them, is it?

benjaoming commented:
> I'm a little confused on the differences in these 2 pages. I think a better title on Configuring environment variables would help. It's not about configuring them, is it?

It was a terrible title!

There are currently 3 pages in play, all of which are largely finished methinks..

1) A how-to about defining environment variables for a project
2) A reference to built-in environment variables that are available in the build process
3) An explanation about accessing environment variables in the build process

The "3rd page" was called "Configuring environment variables" and the title wasn't good. I've changed it to "Understanding environment variables" and added a better introduction and some more content about sphinx-multiproject, since that AFAICT is a very solid reason to mention here.

ericholscher commented on docs/user/reference/environment-variables.rst:
```diff
@@ -0,0 +1,50 @@
+
+    :Examples: ``en``, ``it``, ``de_AT``, ``es``, ``pt_BR``
+
+.. envvar:: READTHEDOCS_PROJECT
```
I'd put this above language too, since it relates to the project. I always think about them as the order in the URL:

`<project>.readthedocs.io/<language>/<version>/`

ericholscher commented on docs/user/reference/environment-variables.rst:
```diff
@@ -0,0 +1,50 @@
+    Path for the :ref:`virtualenv that was created for this build <builds:Understanding what's going on>`.
+    Only exists for builds using Virtualenv and not Conda.
+
+    :Example: ``/home/docs/checkouts/readthedocs.org/user_builds/project/envs/version``
```
I'd throw a `seealso` here to the other 2 pages.

benjaoming replied:
Is done!

ericholscher commented on docs/user/guides/environment-variables.rst:
```diff
@@ -0,0 +1,87 @@
+
+   .. warning::
+
+      If you mark this option, any user that can create a pull request
```
```suggestion
      If you make your environment variable **public**, any user that can create a pull request
```

ericholscher commented on docs/user/guides/environment-variables.rst:
```diff
@@ -0,0 +1,87 @@
+      If you mark this option, any user that can create a pull request
+      on your repository will be able to see the value of this environment variable.
+      In other words,
+      do not use this option if your environment variable is a secret.
```
```suggestion
      **do not use this option if your environment variable is a secret.**
```

ericholscher commented on docs/user/guides/environment-variables.rst:
```diff
@@ -0,0 +1,87 @@
+      In other words,
+      do not use this option if your environment variable is a secret.
+
+.. note::
```
Don't love 2 callouts back to back, but seems OK in this case? Not sure this one needs a note vs. just normal text.

benjaoming replied:
It was fairly easy to convert it to normal text, yes :+1:

ericholscher commented on docs/user/guides/environment-variables.rst:
```diff
@@ -0,0 +1,87 @@
+
+
+Finally, click on :guilabel:`Save`.
+Your custom environment variable is immediately active for all future builds
+and you are ready to start using it.
```
💯

ericholscher commented on docs/user/guides/environment-variables.rst:
```diff
@@ -0,0 +1,87 @@
+Your custom environment variable is immediately active for all future builds
+and you are ready to start using it.
+
+In the next sections,
+you can see a few examples of using environment variables. ⬇️
```
```suggestion
Keep reading for a few examples of using environment variables. ⬇️
```

`next sections` is awkward.

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
-* User-defined variables are not available during the checkout step of the :doc:`build process </builds>`
-* User-defined variables that are not marked as public will not be available in :doc:`pull request builds </pull-requests>`
+Environment variables are defined in the :term:`dashboard` interface in :menuselection:`Admin --> Environment variables`.
+Environment variables are configured and managed for a project's entire build process `with 2 exceptions <Environment variables and build environments>`_.
```
I think if we use a `:ref:` here it will hover :) Also the link is broken. I believe `:ref:` will also fail the build on broken links. We should almost certainly fail the build for broken _internal_ links as well.

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
------------------------------
+Aside from storing secrets,
+there are :ref:`other patterns <environment-variables:Patterns of using environment variables>` that take advantage of environment variables,
+for instance to reuse the same *monorepo* configuration in multiple documentation projects.
```
```suggestion
like reusing the same *monorepo* configuration in multiple documentation projects.
```

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
+Aside from storing secrets,
+there are :ref:`other patterns <environment-variables:Patterns of using environment variables>` that take advantage of environment variables,
+for instance to reuse the same *monorepo* configuration in multiple documentation projects.
+In cases where the environment variable isn't a secret,
```
```suggestion
In cases where the environment variable isn't a secret,
like a build tool flag,
```

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
 
-* `Default environment variables`_
-* `User-defined environment variables`_
+Read the Docs allows you to define your own environment variables to be used in the build process.
+This is useful for adding build secrets such as API tokens.
```
```suggestion
Read the Docs allows you to define your own environment variables to be used in the build process.
It also defines a set of :doc:`default environment variables</reference/environment-variables>` with information about your build.
These are useful for different purposes:

* Custom environment variables are useful for adding build secrets such as API tokens.
* Default environment variables are useful for varying your build specifically for Read the Docs or specific types of builds on Read the Docs. 
```

ericholscher replied:
I don't feel too strongly. We use `default` in a _lot_ of places (eg. `default` values of things). I think `default` makes sense if they can be overwritten, but I don't think that's common here, so `pre-defined` is probably better. Or even `Build-specific environment variables` to be even more explicit and useful (who cares if they are pre-defined, what _are_ they?)

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
 
-* User-defined variables are not available during the checkout step of the :doc:`build process </builds>`
-* User-defined variables that are not marked as public will not be available in :doc:`pull request builds </pull-requests>`
+Environment variables are defined in the :term:`dashboard` interface in :menuselection:`Admin --> Environment variables`.
```
```suggestion
Custom environment variables are defined in the :term:`dashboard` interface in :menuselection:`Admin --> Environment variables`.
```

benjaoming replied:
Yes please!!! :pray:

ericholscher commented on docs/user/environment-variables.rst:
```diff
@@ -1,116 +1,127 @@
-* User-defined variables are not available during the checkout step of the :doc:`build process </builds>`
-* User-defined variables that are not marked as public will not be available in :doc:`pull request builds </pull-requests>`
+Environment variables are defined in the :term:`dashboard` interface in :menuselection:`Admin --> Environment variables`.
+Environment variables are configured and managed for a project's entire build process `with 2 exceptions <Environment variables and build environments>`_.
```
```suggestion
Environment variables are defined for a project's entire build process,
`with 2 important exceptions <Environment variables and build environments>`_.
```

benjaoming replied:
Thanks! Yes! Gonna fix the link as well...

ericholscher reviewed (approved):
This is a 💯 refactor. Thanks for explaining it.
